nmpolicy/internal/resolver/resolver.go

```diff
@@ -0,0 +1,58 @@
+package resolver
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
+)
+
+type Resolver struct {
+	CurrentState map[string]interface{}
+}
+
+func (c *Resolver) Resolve(rootNode ast.Node) (map[string]interface{}, error) {
+	if rootNode.EqFilter != nil {
+		inputSource, err := c.getInputSource(rootNode.EqFilter[0])
+		if err != nil {
+			return nil, resolveError(fmt.Errorf("failed to discover the input source : %v", err))
+		}
+
+		path, err := c.getPath(rootNode.EqFilter[1])
+		if err != nil {
+			return nil, resolveError(fmt.Errorf("failed to discover the path : %v", err))
+		}
+		filteredValue, err := c.filteredValue(rootNode.EqFilter[2])
+		if err != nil {
+			return nil, resolveError(fmt.Errorf("failed to discover the filtered value : %v", err))
+		}
+
+		return filter(inputSource, *path, *filteredValue)
+	}
+	return nil, resolveError(fmt.Errorf("root node has unsupported operation : %v", rootNode))
+}
+
+func (c *Resolver) getInputSource(inputSourceNode ast.Node) (map[string]interface{}, error) {
+	//if inputSourceNode.Identity != nil && *inputSourceNode.Identity == ast.CurrentStateLiteral {
+	if reflect.DeepEqual(inputSourceNode.Terminal, ast.CurrentStateIdentity()) {
+		return c.CurrentState, nil
+	}
+
+	return nil, fmt.Errorf("Not supported input source %v. Only the current state is supported", inputSourceNode)
+}
+
+func (c *Resolver) getPath(pathNode ast.Node) (*ast.VariadicOperator, error) {
+	if pathNode.Path != nil {
+		return pathNode.Path, nil
+	}
+
+	return nil, fmt.Errorf("Invalid path type %T.", pathNode)
+}
+
+func (c *Resolver) filteredValue(filteredValueNode ast.Node) (*ast.Node, error) {
+	if filteredValueNode.String != nil {
+		return &filteredValueNode, nil
+	}
+
+	return nil, fmt.Errorf("Not supported filtered value %v. Only string is supported", filteredValueNode)
+}
```

Comment:
Usually at golang you do the opposite check and return the error kind of "fail first" then after that you put the logic, so the main logic is not nested just the error return.

Reply:
done